### PR TITLE
[FlexibleHeader] Expose a shiftedOffscreen property.

### PR DESCRIPTION
**Pull request by jverkoey**
This property makes it possible to evaluate whether the flexible header has been intentionally hidden or not. This property will be used in a subsequent PR to implement navigationBarHidden in MDCAppBarNavigationController.

Pre-work for https://github.com/material-components/material-components-ios/issues/5185

### CODE REVIEW

**Comment by jverkoey** [comment]
Breakage being addressed in https://github.com/material-components/material-components-ios/pull/9015.

File: components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h [hunk 1]
```diff
@@ -129,6 +129,9 @@ typedef NS_ENUM(NSInteger, MDCFlexibleHeaderContentImportance) {
 /** Asks the receiver to take the header off-screen if it's currently on-screen. */
 - (void)shiftHeaderOffScreenAnimated:(BOOL)animated;
 
+/** Indicates whether the header is presently shifted offscreen. */
+@property(nonatomic, readonly) BOOL shiftedOffscreen;
```
Comment (romoore) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h]: Tip: Consider naming the getter, `isShiftedOffscreen`.

File: components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h [hunk 2]
```diff
@@ -129,6 +129,9 @@ typedef NS_ENUM(NSInteger, MDCFlexibleHeaderContentImportance) {
 /** Asks the receiver to take the header off-screen if it's currently on-screen. */
 - (void)shiftHeaderOffScreenAnimated:(BOOL)animated;
 
+/** Indicates whether the header is presently shifted offscreen. */
```
Comment (romoore) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h]: At what point does it shift from "No" to "yes" and back? Is it if any single pixel of the header is "off-screen" or if all of it is off screen or some significant portion (10%)?

File: components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+  // When
+  [fhvc shiftHeaderOffScreenAnimated:NO];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), -fhvc.maximumHeight, 0.001);
+}
+
+- (void)testBecomesTrueWhenShiftedOffscreenAnimatedWithoutTrackingScrollView {
```
Comment (romoore) [inline components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m]: Shouldn't this test wait until the animation finishes? Or should the test be named something like "becomesTrueBeforeAnimationWhen..."?

File: components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  // When
+  scrollView.contentOffset = CGPointMake(0, 400);
+  [fhvc trackingScrollViewDidScroll];
+
+  // Then
+  XCTAssertTrue(fhvc.shiftedOffscreen);
+  XCTAssertEqualWithAccuracy(CGRectGetMinY(fhvc.frame), -fhvc.maximumHeight, 0.001);
+}
+
+- (void)testStaysFalseWhenDraggedPartiallyOffscreen {
```
Comment (romoore) [inline components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m]: Consider adding a test case (or pair of cases) right at the boundary condition of yes/no for `.shiftedOffscreen` to verify the boundary behaves as expected.  I'm especially curious to understand if the property "waits" to flip until significant change is detected.

For example, imagine the boundary is at 10% with a 5% "hysteresis" margin to avoid rapid flipping:
shifted = NO, offset = 0
offset = 10% of height, shifted = NO
offset = 15% of height, shifted = YES
offset = 10% of height, shifted = YES // hysteresis effect to avoid "flipping" too often at boundary conditions
offset = 5% of height, shifted = NO // Finally flips because exceeded margin beyond trigger point

**Review by romoore — APPROVED** [review]
LGTM, but would recommend more detailed documentation about exactly when the property changes, since it seems to permit some small amount of shifting before it's considered "off screen".

File: components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h (hunk 1 above)
Comment (jverkoey) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h]: Done.

File: components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h (hunk 2 above)
Comment (jverkoey) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView+ShiftBehavior.h]: Clarified the docs.

File: components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m (hunk 3 above)
Comment (jverkoey) [reply components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m]: Renamed the test.

File: components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m (hunk 4 above)
Comment (jverkoey) [reply components/FlexibleHeader/tests/unit/FlexibleHeaderShiftedOffscreenTests.m]: Done. I added `testBecomesFalseWhenScrolledGreaterThanTheHeightOfTheHeaderAndThenBackToTheContentTop` to test going beyond the flipping of the state and then back again.
